crates/memory/src/region.rs

```diff
@@ -7,10 +7,10 @@ pub struct Region {
 impl Region {
     pub fn overlaps(&self, rhs: Region) -> bool {
         let self_start = self.start as u64;
-        let self_end = self.start as u64 + self.len as u64;
+        let self_end = ((self.start + self.len) as i64 - 1) as u64;
 
         let rhs_start = rhs.start as u64;
-        let rhs_end = rhs.start as u64 + rhs.len as u64;
+        let rhs_end = ((rhs.start + rhs.len) as i64 - 1) as u64;
```

Comment:
This seems to be a tiny bug in the pointer arithmetic since `end = len - 1` rather than `end = len` in which case we get the actual `len' = len + 1`.

Reply:
Good catch. We still want to up-cast the u32s to u64 before adding, to eliminate any possibility of overflow.

Reply:
OK, this is now done in e5fd4e6. I'm wondering, should we ideally make this fallible? I'm not necessarily thinking of doing it right now at this very instance; rather, wondering for the future. If there is an overflow or whatnot, return an error or worse, panic. What do you reckon?

Reply:
This code should be infallible - the question of whether one memory region overlaps with another is true or false (even modulo changes like base system word size or wasm word size, if you assume wasm64 will be real one day), it shouldn't be possible to select inputs to that which cause a panic. Since untrusted code will get to select one of the inputs, we will have to fuzz all of these interfaces eventually.

Reply:
Makes sense, thanks!